speccyMac/Loaders/Z80Loader.swift

```diff
@@ -16,6 +40,152 @@ class Z80Loader: GameLoaderProtocol {
+            if data.count > (48 * 1024) + v1HeaderLen {
+                print(" file too long")
+                return false
+            }
+
+            let isCompressed = v1.flags1 & (1 << 5) != 0
+            print("  \(!isCompressed ? "un" : "")compressed data")
+
+
+            // prepare CPU and hardware
+
+            z80.i = v1.i
+
+            z80.exhl = v1.exhl
+            z80.exde = v1.exde
+            z80.exbc = v1.exbc
+            z80.exaf = UInt16(v1.exa) + (UInt16(v1.exf) << 8)
+
+            z80.hl.value = v1.hl
+            z80.de.value = v1.de
+            z80.bc.value = v1.bc
+            z80.iy = v1.iy
+            z80.ix = v1.ix
+
+            // sna[19].00010000 = z80[29].00000001
+            if v1.flags2 & 0x04 > 0 {
+                z80.interrupts = true
+                z80.iff1 = 1
+                z80.iff2 = 1
+            } else {
+                z80.interrupts = false
+                z80.iff1 = 0
+                z80.iff2 = 0
+            }
+
+            z80.r.value = v1.r
+
+            z80.af.value = UInt16(v1.a) + (UInt16(v1.f) << 8)
+            ZilogZ80.sp = v1.sp
+
+            // sna[25].00000011 = z80[29].00000011
+            z80.interruptMode = v1.flags2 & 0x03
+            // sna[26].00000111 = z80[12].00001110
+            z80.machine?.output(0xfe, byte: (v1.flags1 >> 1) & 0x07)
+
+            // prepare ram
+
+            let ram = z80.memory.romSize
+            let ramlen = 64 * 1024 // can change if 16k support is added
+            let datlen: UInt16 = UInt16(data.count) - v1HeaderLen
+            var i: UInt16 = v1HeaderLen
+            var o: UInt16 = 0
+
+            var loop = true
+            while loop {
+                let left = datlen - i // TODO can overflow?
+                if left == 0 {
+                    print("    end of data")
+                    break
+                }
+                if UInt(ram) + UInt(o) >= ramlen {
+                    print("    end of ram")
+                    return false
+                }
+
+                let b = data[Int(i)]
+
+                if isCompressed && b == 0x00 && left >= 4 && data[Int(i + 1)] == 0xed
+                    && data[Int(i + 2)] == 0xed && data[Int(i + 3)] == 0x00 {
+                    print("    data terminator")
+                    loop = false
+                    i += 4
+                } else if isCompressed && b == 0xed && left >= 4 && data[Int(i + 1)] == 0xed {
+                    let n = data[Int(i + 2)]
+                    let v = data[Int(i + 3)]
+
+                    for _ in 0..<n {
+                        z80.memory.set(ram + o, byte:v)
+
+                        o += 1
+                    }
+                    i += 4
+                } else {
+                    z80.memory.set(ram + o, byte:b)
+
+                    i += 1
+                    o += 1
+                }
+            }
+
+            // start CPU
+
+            let lo = z80.memory.get(ZilogZ80.sp)
```

Comment:
I don't think you need to do this for a .z80 file - isn't this specific to .sna? (should be able to remove lines 172 to 182, just need to return true now, you've already set PC

Reply:
Ah I just copied your code thinking that was necessary, but I did stare at it a few times wondering. I've written code to analyse the formats before but not as part of an emulator till now. This could be why quite a few .z80s lock up or crash on loading.

Reply:
Actually when I delete those lines, Bruce Lee now stops on the loading screen, but works sometimes depending on which game I was running just prior. Technician Ted now crashes unless I was running the .sna version just prior.

It's 1.30am here so I won't be able to reason about it until tomorrow.

Reply:
Actually I hadn't set the PC. You probably saw where I got it from the `Data` into the `v1` header struct. So I thought just adding `z80.pc = v1.pc` at line 172 would fix everything. But while it makes Technician Ted work perfectly, Bruce Lee still locks up.